main.py

```
@@ -25,5 +25,50 @@ async def build_esp_related_project(project_path: str) -> (str, str):
+@mcp.tool()
+async def list_esp_serial_ports() -> (str, str):
+    """List available serial ports for ESP devices.
+
+    Returns:
+        tuple: (stdout, stderr) - Available serial ports and any error messages
+    """
+    returncode, stdout, stderr = await list_serial_ports()
+    
+    # Log the port listing operation
+    port_log = f"Port listing - Return code: {returncode}\nSTDOUT:\n{stdout}\nSTDERR:\n{stderr}"
+    open('mcp-ports.log', 'w+').write(port_log)
```

Ruff (0.11.9)

Use a context manager for opening files

(SIM115)